Semantic tokens from LS available in the Java editor

BoykoAlex opened the pull request:
Based on the latest addition of semantic tokens extension for Java editor the support is added for having semantic tokens from the language server in the Java editor.
There is a preference add to turn off semantic tokens support in the Java editor. The preference is defined but the UI is not and is expected to come from 3rd party plugins using this functionality (i.e. spring boot ls plugin) otherwise it's unclear where to fit this preference in the UI at the moment

rubenporras commented on org.eclipse.lsp4e.jdt/.settings/org.eclipse.jdt.core.prefs:
```diff
@@ -9,7 +9,7 @@ org.eclipse.jdt.core.compiler.annotation.nonnullbydefault.secondary=
 org.eclipse.jdt.core.compiler.annotation.nullable=org.eclipse.jdt.annotation.Nullable
 org.eclipse.jdt.core.compiler.annotation.nullable.secondary=
-org.eclipse.jdt.core.compiler.annotation.nullanalysis=enabled
+org.eclipse.jdt.core.compiler.annotation.nullanalysis=disabled
```
would you not prefer to follow the other projects in this settings? What is the reason for disabling it?

BoykoAlex replied:
I would... but I'm lost with JDT Null check annotations... I see that LSP4E plugin project has lots of errors coming from JDT related to Null checks but the maven build goes through. The `lsp4e.jdt` project also has a bunch of same origin errors however maven build fails with compilation errors for the JDT null check issues... I wasn't able to figure out where the difference between `lsp4e` and `lsp4e.jdt` projects comes from... Disabling the null checking in the preference solved it for the `lsp4e.jdt` for me...

BoykoAlex replied:
It seems that I need to add `@NonNull` annotations in JDT UI in `ISemanticTokensProvider` and the plugin there depend on `org.eclipse.jdt.annotation` plugin... Otherwise without ignoring null violation analysis i can't get it to compile :-\

rubenporras replied:
Sorry for the long delay.

You can annotate libraries externally with annotations, in LSP4E we use the library https://github.com/vegardit/no-npe instead of doing it directly in our repo. You should be able to fix it by adding annotations in https://github.com/vegardit/no-npe/tree/main/libs/eea-eclipse-platform/src/main/resources/org/eclipse

BoykoAlex replied:
Hmm... I didn't see `eaa-lsp4e` in there... Or did you mean I should add `lsp4e` in there?

rubenporras commented on org.eclipse.lsp4e.jdt/src/org/eclipse/lsp4e/jdt/internal/LSJavaHoverProvider.java:
```diff
@@ -9,7 +9,7 @@
  *  - Martin Lippert (Pivotal Inc.) - Initial implementation
  *******************************************************************************/
-package org.eclipse.lsp4e.jdt;
+package org.eclipse.lsp4e.jdt.internal;
```
shall we not rather do this move in different PR?

BoykoAlex replied:
Yes, we can do that. I'll revert moving classes. I did that because had to export a package and didn't mean to export more than needed... Wonder if perhaps we could have a preference page for LSP4E JDT bit and expose the preference I'm introducing on that new preference page... This way we can completely avoid exposing anything as API...

rubenporras commented on target-platforms/target-platform-latest/target-platform-latest.target:
```diff
@@ -13,7 +13,7 @@
       <unit id="org.eclipse.ui.tests.harness" version="0.0.0"/>
       <unit id="org.eclipse.sdk.ide" version="0.0.0"/>
-      <repository location="https://download.eclipse.org/eclipse/updates/latest/"/>
+      <repository location="https://download.eclipse.org/eclipse/updates/4.35-I-builds/"/>
```
Am I correct in guessing that this change means that the PR cannot be merged before 4.35 has been released?

BoykoAlex replied:
Yes, it'd need 4.35. Are LSP4E releases come out as new Eclipse being released? When does LSP4E adopt changes in the platform? When the new Eclipse is out already? Or a few months/weeks bore new Eclipse is out?

mickaelistria commented on org.eclipse.lsp4e.jdt/META-INF/MANIFEST.MF:
```diff
@@ -14,4 +17,9 @@ Require-Bundle: org.eclipse.core.runtime;bundle-version="3.12.0",
  org.eclipse.lsp4e;bundle-version="0.18.13",
  org.eclipse.jdt.core;bundle-version="3.20.0",
  org.eclipse.jdt.ui;bundle-version="3.20.0",
```
The version should be bumped to the 1st version that introduces `org.eclipse.jdt.ui.text.java.ISemanticTokensProvider`

BoykoAlex commented:
@rubenporras I've rebased this PR on the latest from main however I still get the pesky Null Annotations compiler error:
```
  Error: 6,958 [ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:4.0.12:compile (default-compile) on project org.eclipse.lsp4e.jdt: Compilation failure: Compilation failure: 
  Error: 6,958 [ERROR] /Users/runner/work/lsp4e/lsp4e/org.eclipse.lsp4e.jdt/src/org/eclipse/lsp4e/jdt/LSJavaSemanticTokensProvider.java:[42] 
  Error: 6,958 [ERROR] 	public Collection<ISemanticTokensProvider.SemanticToken> computeSemanticTokens(CompilationUnit ast) {
  Error: 6,958 [ERROR] 	       ^^^^^^^^^^
  Error: 6,958 [ERROR] The return type is incompatible with 'Collection<ISemanticTokensProvider.SemanticToken>' returned from ISemanticTokensProvider.computeSemanticTokens(CompilationUnit) (mismatching null constraints)
  Error: 6,958 [ERROR] 1 problem (1 error)
  Error: 6,958 [ERROR] -> [Help 1]
```

I have added JDT UI into eclipse platform with the new `ISemanticTokensProvider` API: https://github.com/vegardit/no-npe/blob/main/libs/eea-eclipse-platform/src/main/resources/org/eclipse/jdt/ui/text/java/ISemanticTokensProvider.eea but it doesn't seem to help :-\. Is there anything else I can do other than change the preferences for LSP4E JDT plugin to flag this issues as a warning rather than error?

mickaelistria commented:
@sebthom Can you please help regarding this null analysis error?

sebthom commented on org.eclipse.lsp4e.jdt/src/org/eclipse/lsp4e/jdt/JavaSemanticTokensProcessor.java:
```diff
@@ -0,0 +1,116 @@
+		}
+	}
+
+	private List<String> tokenModifiers(final Integer data, final List<String> legend) {
```
This method is unused code and should be removed.

sebthom commented on org.eclipse.lsp4e.jdt/src/org/eclipse/lsp4e/jdt/JavaSemanticTokensProcessor.java:
```diff
@@ -0,0 +1,116 @@
+	 * @param semanticTokensLegend
+	 * @return
+	 */
+	public @NonNull List<ISemanticTokensProvider.SemanticToken> getSemanticTokens(@NonNull final List<Integer> dataStream,
```
the NonNull annotations are redundant since non-null is enabled by default for this package in package-info.java

sebthom commented:
> I have added JDT UI into eclipse platform with the new `ISemanticTokensProvider` API: [vegardit/no-npe@`main`/libs/eea-eclipse-platform/src/main/resources/org/eclipse/jdt/ui/text/java/ISemanticTokensProvider.eea](https://github.com/vegardit/no-npe/blob/main/libs/eea-eclipse-platform/src/main/resources/org/eclipse/jdt/ui/text/java/ISemanticTokensProvider.eea?rgh-link-date=2025-03-31T17%3A14%3A40Z) but it doesn't seem to help :-. Is there anything else I can do other than change the preferences for LSP4E JDT plugin to flag this issues as a warning rather than error?

I can reproduce this. It is a limitation in the Eclipse compiler's null analysis. In this case the `@NonNullByDefault` annotation in the package-info.java provokes a conflict with the under-specified non-null external annotations of `ISemanticTokensProvider#computeSemanticTokens`. I updated the external annotations file. Please rebase your PR. It should compile now.

BoykoAlex commented:
@sebthom many thanks for helping with Null Annotations issue and a code review. Looks like everything is in order to merge this PR now!-)

sebthom commented:
> The preference is defined but the UI is not and is expected to come from 3rd party plugins using this functionality (i.e. spring boot ls plugin) otherwise it's unclear where to fit this preference in the UI at the moment

I haven't looked at the functionality provided by this PR, but is it reasonable to expect third parties to globally enable/disable this, given there's no UI option to turn it off? Considering potential side effects, wouldn’t it make more sense to have the feature disabled by default (and expect 3rd party plugins to enable it) or to directly provide a UI option to enable/disable it?

BoykoAlex commented:
@sebthom I was wondering about the same: https://github.com/eclipse-lsp4e/lsp4e/pull/1161#discussion_r1884173476
If there is a place in the UI where i can put the preference I'd be happy to do so.

rubenporras commented on org.eclipse.lsp4e.jdt/src/org/eclipse/lsp4e/jdt/JavaSemanticTokensProcessor.java:
```diff
@@ -0,0 +1,95 @@
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.SemanticTokensLegend;
+
+class JavaSemanticTokensProcessor {
```
This class has a lot of common code with `org.eclipse.lsp4e.operations.semanticTokens.SemanticTokensDataStreamProcessor`, could we have a common implementation?

rubenporras commented on org.eclipse.lsp4e.jdt/src/org/eclipse/lsp4e/jdt/LSJavaSemanticTokensProvider.java:
```diff
@@ -0,0 +1,162 @@
+	}
+
+	private ISemanticTokensProvider.TokenType mapToTokenType(String tokeTypeStr) {
+		switch (tokeTypeStr) {
```
Could we use the syntax
return switch (tokenTypeStr) {
  case "method" -> ISemanticTokensProvider.TokenType.METHOD
  ...

rubenporras commented on org.eclipse.lsp4e.jdt/src/org/eclipse/lsp4e/jdt/LSJavaSemanticTokensProvider.java:
```diff
@@ -0,0 +1,162 @@
+					return ls.getTextDocumentService().semanticTokensFull(params)
+							.thenApply(semanticTokens -> {
+								if (semanticTokens == null) {
+									return Collections.<SemanticToken>emptyList();
```
```suggestion
									return Collections.emptyList();
```

rubenporras commented on org.eclipse.lsp4e.jdt/src/org/eclipse/lsp4e/jdt/LSJavaSemanticTokensProvider.java:
```diff
@@ -0,0 +1,162 @@
+								}
+								SemanticTokensLegend legend = getSemanticTokensLegend(w);
+								if (legend == null) {
+									return Collections.<SemanticToken>emptyList();
```
```suggestion
									return Collections.emptyList();
```

rubenporras commented on org.eclipse.lsp4e.jdt/src/org/eclipse/lsp4e/jdt/LSJavaSemanticTokensProvider.java:
```diff
@@ -0,0 +1,162 @@
+		return null;
+	}
+
+	private @Nullable SemanticTokensLegend getSemanticTokensLegend(final LanguageServerWrapper wrapper) {
```
This method (and others) is an exact copy of org.eclipse.lsp4e.operations.semanticTokens.SemanticHighlightReconcilerStrategy.getSemanticTokensLegend(LanguageServerWrapper).

Others are very similar.  Do you think we could refactor the code to avoid the redundancy?

sebthom commented on org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/semanticTokens/LspSemanticTokensDataStreamProcessor.java:
Could we call this class AbstractSemanticTokensDataStreamProcessor?

BoykoAlex commented:
@sebthom @rubenporras Thank you for the review comments. Indeed the code needed quite bit of tidy up! Hopefully looks better now. The build fails because there is something off with the HighlightJS style fetching... host server is down it seems...
Anything you'd like me to try with the preference switching on/off LSP semantic tokens in the Java editor? A Java page under "Language Servers" preferences?

sebthom commented:
> Anything you'd like me to try with the preference switching on/off LSP semantic tokens in the Java editor? A Java page under "Language Servers" preferences?

@BoykoAlex @rubenporras  Maybe it would make sense to add a "Features" preferences page under "Language Servers" where one can enable/disable semantic tokens and in future other features.

BoykoAlex commented:
I have added the **Java** preference page under Language Servers preferences. It's contributed via the LSP4E JDT plugin hence would only be shown for those having the LSP4E JDT plugin...

BoykoAlex commented:
Rebased and squashed commits

BoykoAlex commented:
Here is how it looks at the moment. SQL queries are styled after Java editor semantic tokens... i.e. keywords, variables, properties, methods etc.

![Screenshot 2025-04-04 at 16 50 01](https://github.com/user-attachments/assets/6b1f8970-430d-45b9-af4f-8eed45dd8d2b)

BoykoAlex commented:
@rubenporras @sebthom shall we fold this in? More feedback on the PR is welcomed...

rubenporras commented on org.eclipse.lsp4e.jdt/src/org/eclipse/lsp4e/jdt/LSJavaSemanticTokensProvider.java:
```diff
@@ -0,0 +1,107 @@
+			return SemanticHighlightReconcilerStrategy.requestFullSemanticTokens(theDocument, (legend, semanticTokens) -> convertTokens(legend, theDocument, semanticTokens))
+				.thenApply(o -> o.orElse(Collections.emptyList())).get(300, TimeUnit.MILLISECONDS);
+		} catch (InterruptedException | ExecutionException | TimeoutException e) {
+			LanguageServerPlugin.logError("Failed to fetch semantic tokens for '%s' from Language Servers".formatted(resource.getLocation()), e);
```
So far we log `TimeoutException` as a warning that includes the timeout value, could we do that?

rubenporras commented on org.eclipse.lsp4e.jdt/src/org/eclipse/lsp4e/jdt/LSJavaSemanticTokensProvider.java:
```diff
@@ -0,0 +1,107 @@
+		try {
+			return SemanticHighlightReconcilerStrategy.requestFullSemanticTokens(theDocument, (legend, semanticTokens) -> convertTokens(legend, theDocument, semanticTokens))
+				.thenApply(o -> o.orElse(Collections.emptyList())).get(300, TimeUnit.MILLISECONDS);
+		} catch (InterruptedException | ExecutionException | TimeoutException e) {
```
Should we not call `Thread.currentThread().interrupt()` for `InterruptedException`?

BoykoAlex replied:
I suppose we should. Will make this change.

rubenporras commented on org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/semanticTokens/SemanticTokensDataStreamProcessor.java:
```diff
@@ -21,18 +18,14 @@
-
-	private final Function<Position, Integer> offsetMapper;
-	private final Function<String, @Nullable IToken> tokenTypeMapper;
+public class SemanticTokensDataStreamProcessor extends AbstractcSemanticTokensDataStreamProcessor<IToken, StyleRange>{
```
```suggestion
public class SemanticTokensDataStreamProcessor extends AbstractcSemanticTokensDataStreamProcessor<IToken, StyleRange> {
```

rubenporras commented on org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/semanticTokens/SemanticHighlightReconcilerStrategy.java:
```diff
@@ -190,7 +193,7 @@ private void saveStyle(final Pair<@Nullable SemanticTokens, @Nullable SemanticTo
 		final var styleRangeHolder = this.styleRangeHolder;
 		if (!dataStream.isEmpty() && semanticTokensDataStreamProcessor != null && styleRangeHolder != null) {
-			List<StyleRange> styleRanges = semanticTokensDataStreamProcessor.getStyleRanges(dataStream,
+			List<StyleRange> styleRanges = semanticTokensDataStreamProcessor.getTokens(dataStream,
```
to call `getTokens` and to save it into a variable of type `StyleRange` is a bit awkward, but I do not have a suggestion for a better name at this moment, do you have a better idea?

BoykoAlex replied:
`getVisualTokens()`?

rubenporras replied:
since it is an `...TokensDataStreamProcessor`, maybe `getTokensData`?

After that I would be happy to merge.

BoykoAlex replied:
Sounds good to me - made this change.

rubenporras commented on org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/semanticTokens/SemanticHighlightReconcilerStrategy.java:
```diff
@@ -171,7 +174,7 @@ private Function<Position, Integer> offsetMapper() {
 	}
 
-	private SemanticTokensParams getSemanticTokensParams() {
+	private static SemanticTokensParams getSemanticTokensParams(IDocument document) {
```
I see that the `document` is still a class field, even though many methods are now static and take it as an argument. Should we maybe split the class so that we do not have methods which take a variable as an argument that hides a field in the class?

BoykoAlex replied:
I see, let me look into this...

rubenporras commented:
@BoykoAlex , thanks for the reminder, I was very busy the last days and completely forgot this PR. A have made a new round of comments, all but the last one are very minor, thanks for the changes!

BoykoAlex commented:
@rubenporras I addressed your comments with the exception of the method renaming, i.e. `getTokens()`... `getTokenVisuals()`? Not really sure myself of a proper name... `getTokensVisualData()`?

BoykoAlex commented:
Made requested changes and squashed commits

rubenporras reviewed (approved):
thanks
